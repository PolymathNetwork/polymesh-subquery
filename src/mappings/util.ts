import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
