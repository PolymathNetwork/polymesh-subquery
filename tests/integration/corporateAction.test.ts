import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('withholdingTaxesOfCas', () => {
  test('should return the witholding taxes', async () => {
    const q = {
      variables: { ticker: '13TICKER' },
      query: gql`
        query q($ticker: String!) {
          withholdingTaxesOfCas(filter: { ticker: { equalTo: $ticker }, localId: { equalTo: 0 } }) {
            nodes {
              taxes
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});

describe('historyofPaymentEventsForCas', () => {
  test('should get payment history', async () => {
    const q = {
      variables: { ticker: '13TICKER' },
      query: gql`
        query q($ticker: String!) {
          historyOfPaymentEventsForCas(
            filter: { ticker: { equalTo: $ticker }, localId: { equalTo: 0 } }
          ) {
            totalCount
            nodes {
              balance
              blockId
              datetime
              eventDid
              eventId
              eventIdx
              localId
              tax
              ticker
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();

    expect(subquery?.data).toMatchSnapshot();
  });
});
