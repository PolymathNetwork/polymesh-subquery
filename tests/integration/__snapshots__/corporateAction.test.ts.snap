// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`historyofPaymentEventsForCas should get payment history 1`] = `
Object {
  "historyOfPaymentEventsForCas": Object {
    "__typename": "HistoryOfPaymentEventsForCasConnection",
    "nodes": Array [
      Object {
        "__typename": "HistoryOfPaymentEventsForCa",
        "balance": 100000000,
        "blockId": 358,
        "datetime": "2021-10-06T12:35:04",
        "eventDid": "0x6c15581b6ded46773fbb9786b44624fb2d44397f89bfd71b077462c22ac4b72d",
        "eventId": "BenefitClaimed",
        "eventIdx": 2,
        "localId": 0,
        "tax": 15,
        "ticker": "13TICKER",
      },
    ],
    "totalCount": 1,
  },
}
`;

exports[`withholdingTaxesOfCas should return the witholding taxes 1`] = `
Object {
  "withholdingTaxesOfCas": Object {
    "__typename": "WithholdingTaxesOfCasConnection",
    "nodes": Array [
      Object {
        "__typename": "WithholdingTaxesOfCa",
        "taxes": 1500,
      },
    ],
  },
}
`;
